### PR TITLE
Add ActivityContext parsing APIs

### LINKED ISSUE #40046: System.Diagnostics.ActivitySource API Addition
## Background and Motivation

I'm working on implementing the new System.Diagnostics.ActivitySource API using the OpenTelemetry-dotnet beta that was just released on top of it. I've run into a couple of pain points that I think an API tweak or two would help alleviate.

I'm consuming messages off a queue. These messages have W3C TraceParent and TraceState values stored as message headers. If I wanted to create an Activity using these values as a parent the API provides:

```csharp
public sealed class ActivitySource
{
  public Activity? StartActivity(string name, ActivityKind kind, string parentId, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default);
}
```

That would work fine but I don't want to create a parented Activity, I want to add this TraceContext as a link on that Activity.

`ActivityLink` only accepts `ActivityContext` in its ctor. There's actually no way I can find to parse a string TraceParent + TraceState into an `ActivityContext`. That functionality is available internally.

## Proposed API

```csharp
namespace System.Diagnostics
{
    public readonly partial struct ActivityContext : IEquatable<ActivityContext>
    {
+        public static bool TryParse(string w3cId, string? traceState, out ActivityContext context);
    }
}
```

What I'm looking for is a way to expose to users the internal parsing so I can turn my strings into a valid `ActivityContext` without a lot of additional work.

## Workaround

What I'm doing right now is just implementing some of the internal logic myself...

```csharp
		public static ActivityContext CreateActivityContextFromW3CTraceContext(string traceParent, string? traceState = null)
		{
			if (string.IsNullOrEmpty(traceParent))
				throw new ArgumentNullException(nameof(traceParent));

			if (traceParent.Length != 55)
				throw new FormatException($"TraceParent [{traceParent}] format is not supported.");

			ReadOnlySpan<char> traceIdSpan = traceParent.AsSpan(3, 32);
			ReadOnlySpan<char> spanIdSpan = traceParent.AsSpan(36, 16);

			return new ActivityContext(
				ActivityTraceId.CreateFromString(traceIdSpan),
				ActivitySpanId.CreateFromString(spanIdSpan),
				(ActivityTraceFlags)ConversionExtensions.HexByteFromChars(traceParent[53], traceParent[54]),
				traceState);
		}
```

/cc @tarekgh @noahfalk @cijothomas @eddynaka @alanwest

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #40046

### CODE REVIEW

**Comment by tarekgh** [comment]
@CodeBlanch @cijothomas @lmolkova

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 1]
```diff
@@ -382,6 +382,29 @@ public void TestDefaultParentContext()
             }).Dispose();
         }
 
+        [Fact]
+        public void TestActivityContextParsing()
+        {
+            const string w3cId = "00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-01";
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:388]: can you add a test passing a non-valid w3cId as input?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs [hunk 2]
```diff
@@ -59,6 +59,40 @@ public ActivityContext(ActivityTraceId traceId, ActivitySpanId spanId, ActivityT
...
+        /// Parse W3C trace parent and state headers to ActivityContext object.
+        /// </summary>
+        /// <param name="traceParent">W3C trace parent header.</param>
+        /// <param name="traceState">Trace state.</param>
+        /// <param name="context">The ActivityContext object created from the parsing operation.</param>
+        public static bool TryParse(string traceParent, string? traceState, out ActivityContext context)
+        {
+            if (traceParent == null)
+            {
+                throw new ArgumentNullException(nameof(traceParent));
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs:72]: Is throwing okay from "tryParse" methods? Should we instead simply return false?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs [hunk 3]
```diff
@@ -59,6 +59,40 @@ public ActivityContext(ActivityTraceId traceId, ActivitySpanId spanId, ActivityT
         /// </remarks>
         public bool IsRemote { get; }
 
+        /// <summary>
+        /// Parse W3C trace parent and state headers to ActivityContext object.
+        /// </summary>
+        /// <param name="traceParent">W3C trace parent header.</param>
+        /// <param name="traceState">Trace state.</param>
```
Comment (CodeBlanch) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs:66]: nits: Kind of odd to have W3C in the one comment but not the other. How about...
```csharp
        /// <param name="traceParent">W3C traceparent header value.</param>
        /// <param name="traceState">W3C tracestate header value.</param>
```
Personally not a fan of the casing "traceparent" & "tracestate" but that's how they are named in the [spec](https://www.w3.org/TR/trace-context/#design-overview).

For the method summary, might want to refer to them collectively as Trace Context? Eg:
```csharp
        /// <summary>
        /// Parse W3C Trace Context header values into an ActivityContext object.
        /// </summary>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs:72]: Yes, all TryParse validate the parameters and throws on nulls.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs]: > Personally not a fan of the casing "traceparent" & "tracestate" but that's how they are named in the spec.

I prefer to case it as the rest of C# method parameters.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:388]: I added a test. Thanks.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs]: Thanks @CodeBlanch I have fixed the comments as you suggested.

**Review by noahfalk — APPROVED** [review]
LGTM 👍

**Comment by tarekgh** [comment]
The failed tests are not related and tracked. merging this one.

Thanks all for your reviews.

**Pull request merged**